net: http: service: enhance HTTP service with optional custom socket create

AndreyDodonov-EH opened the pull request:
Currently webserver does not support custom socket creation e.g. for offloaded socket.

Update the HTTP service API to allow custom socket creation through a function pointer in the service descriptor.
The existing socket creation method is retained as a fallback if no custom function is provided.

This functionality is per-service, i.e. it allows to specify which http services will work on which sockets.
Backwards-compatible.

Partially related to https://github.com/zephyrproject-rtos/zephyr/issues/89960

rlubos commented on include/zephyr/net/http/service.h:
```diff
@@ -220,14 +225,14 @@ struct http_service_desc {
  */
 #define HTTPS_SERVICE_DEFINE(_name, _host, _port, _concurrent, _backlog, _detail,              \
-			     _res_fallback, _sec_tag_list, _sec_tag_list_size)                 \
+			     _res_fallback, _sec_tag_list, _sec_tag_list_size, ...) \
```
TBH I think I'd prefer to keep all of the macro parameters explicit instead of hiding optional parameters behind `__VA_ARGS__` - it might get super confusing if someone wanted to add yet another parameter in the future, it'd be tricky to document also.

We could just extend the macros/service struct with an additional `struct http_service_config` where we could keep any optional per-service configurations, including this custom socket struct, so in case any other custom configurations will be needed in the future, we could just extend the struct w/o touching the macros again.

AndreyDodonov-EH replied:
@rlubos
Good point!
If we want to both prevent future modifications and stay backwards compatible,
I can also alternatively create new macroses which accept `struct http_service_config`

rlubos reviewed (commented):
Looks good, IMHO we can proceed with this approach.

AndreyDodonov-EH commented:
Updated tests and docs.

Also updated release notes, not sure if it is needed for 4.2

kartben commented:
> Updated tests and docs.
> 
> Also updated release notes, not sure if it is needed for 4.2

Please update 4.3 release notes instead given this won't make it into 4.2 which is currently in feature freeze and a week away from being released. Thanks!

AndreyDodonov-EH committed:
net: http: service: extend HTTP service with config

Update the HTTP service API to allow setting per-service configuration,
currently it is only custom socket creation callback,
but in the future it can be extended with other parameter

Signed-off-by: Andrey Dodonov <Andrey.Dodonov@endress.com>

rlubos reviewed (dismissed):
Looks good, thanks!

jukkar commented:
@AndreyDodonov-EH could you check the merge conflict please

AndreyDodonov-EH commented:
> @AndreyDodonov-EH could you check the merge conflict please

@jukkar Done! Thanks for bringing it to attention.

It was just

```
Networking
**********

.. zephyr-keep-sorted-start re(^\w)

.. zephyr-keep-sorted-stop
```

I overwrote it with my documentation entry, since I think sorting is not really applicable in sub-entries, since they always contain `*`, i.e. bullet points.

jukkar commented on doc/releases/migration-guide-4.3.rst:
```diff
@@ -64,9 +64,10 @@ Ethernet
 Networking
 **********
 
-.. zephyr-keep-sorted-start re(^\w)
-
-.. zephyr-keep-sorted-stop
```
The sorted options were just recently added and the other subsystems have those too so I think we should not remove these.

kartben replied:
@pdgendgt

AndreyDodonov-EH replied:
@cfriedt @pdgendt 
How should I fix it?
Looking at recent release notes https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/migration-guide-4.2.rst under Networking there are directly bullet points.
Does the sorting skip it and just uses the letters after that?

So should I do the following? Or what's the expected structure?
```
.. zephyr-keep-sorted-start re(^\w)

* The HTTP server now respects the configured ``_config`` value. Check that
  you provide applicable value to :c:macro:`HTTP_SERVICE_DEFINE_EMPTY`,
  :c:macro:`HTTPS_SERVICE_DEFINE_EMPTY`, :c:macro:`HTTP_SERVICE_DEFINE` and
  :c:macro:`HTTPS_SERVICE_DEFINE`.

.. zephyr-keep-sorted-stop
```

AndreyDodonov-EH replied:
@pdgendt @cfriedt 
Done

cfriedt reviewed (changes requested):
temporary NAK until migration guide is updated

jukkar commented on doc/releases/migration-guide-4.3.rst:
```diff
@@ -64,6 +64,11 @@ Ethernet
 Networking
 **********
 
+* The HTTP server now respects the configured ``_config`` value. Check that
```
I am baffled, I thought that one is suppose to add the needed entries between start and stop directives. @pdgendt, is the usage documented somewhere?

pdgendt replied:
The `re(^\w)` makes it so that the compliance check only verifies that lines starting with a word are sorted. So these bullet points would be skipped anyway.

Documentation is lacking, yes.

AndreyDodonov-EH commented:
> temporary NAK until migration guide is updated

@cfriedt Updated

cfriedt commented:
@AndreyDodonov-EH - it will be merged probably around 5 hours from now. Thanks for the update!
